### ISSUE TITLE
Ensembling forecasters: Checking to see if forecaster has already been fit

### ISSUE THREAD

**Issue by ngupta23** [alan-turing-institute/sktime#584]
**Is your feature request related to a problem? Please describe.**
Currently, the ensembling process [refits all forecasters](https://github.com/alan-turing-institute/sktime/blob/30b06140eb6e94b2845e6239efd679414dc13188/sktime/forecasting/compose/_ensemble.py#L52) which can be slow if they have already been fit before passing to the ensembling process. 

**Describe the solution you'd like**
Add a check to see if the forecaster has already been fit. It it has been fit already, it does not need to be refit again. This will save on training time.

**Comment by mloning** [comment]
Hi @ngupta23, do you want an ensemble for fitted estimators only aggregating the output from `predict`?

**Comment by ngupta23** [comment]
Hi @mloning , Yes. Basically, in my flow I have several models that I have trained individually and I finally want to also ensemble them as well. But during ensembling, they get trained again which seems like duplicated effort. I was wondering if a check could be added to see if the forecasters are already fitted and skip the fit for them to save time. When I looked at the code, it did not seem that the fit in the ensemble was any different from a regular fit so I think this would still work. What do you think?

**Comment by mloning** [comment]
This is an interesting question, I'm not really sure what the best solution is. My intuition is that if we change how the ensemble works, we should also be open to change other composition classes like pipelines and I'm less sure about  that. 

This seems to be deployment/production code. Why don't you fit all of the estimators with the ensemble class and then save the entire fitted ensemble?

**Comment by ngupta23** [comment]
@mloning The issue is that the ensemble does not always produce the best results. So I want to fit the individual models as well and then pick the best from the ensemble/individual models based on temporal CV stats, etc.

**Comment by mloning** [comment]
But you could pass `GridSearchCV` objects to the ensemble, right? So that the individual models are tuned?

**Comment by ngupta23** [comment]
Yes, here is an example of this. The `EnsembleForecaster` would fit these individual model before ensemble (current flow in sktime). 

But I could uncomment the individual model fits (in code below) for Model 1, 2, and 3 (`gscv_arima`, `gscv_ets`, `gscv_expo`) so they are fit separately and then pass those fitted objects to the ensemble. In that case, would just checking the `_is_fitted` method not be enough to skip the fitting again?

This will enable me to do the final comparison of all 4 models instead of simply picking the ensemble by default.

```python
# With Grid Search and with CV (Variant 2)
from sktime.forecasting.arima import AutoARIMA
from sktime.forecasting.ets import AutoETS

cv = SlidingWindowSplitter(
    initial_window=int(len(y_train) * 0.5),
    start_with_window=True,
    window_length=36,
)

# Model 1
forecaster_param_grid = {'sp': [3, 6, 12, 24]}
forecaster_arima = AutoARIMA(suppress_warnings=True)
gscv_arima = ForecastingGridSearchCV(forecaster_arima, cv=cv, param_grid=forecaster_param_grid, verbose=True)
# %time gscv_arima .fit(y_train)

# Model 2
forecaster_param_grid = {
    'sp': [3, 4, 12, 24],    
    }
forecaster_ets = AutoETS(auto=True)
gscv_ets = ForecastingGridSearchCV(forecaster_ets, cv=cv, param_grid=forecaster_param_grid, verbose=True)
# %time gscv_ets.fit(y_train)

# Model 3
forecaster_param_grid = {
    'sp': [3, 4, 12, 24],
    'trend': ["add", "mul", None],
    'damped_trend': [True, False],
    'seasonal': ["add", "mul", None]
    }
forecaster_expo = ExponentialSmoothing()
gscv_expo = ForecastingGridSearchCV(forecaster_expo, cv=cv, param_grid=forecaster_param_grid, verbose=True)
# %time gscv_expo.fit(y_train)

# Model 4
forecaster_ensemble = EnsembleForecaster([
    ("autoARIMA", gscv_arima),
    ("autoETS", gscv_ets),
    ("autoExpo", gscv_expo)
])
%time forecaster_ensemble.fit(y_train)

#####################################################################################
## Finally compare all 4 models (3 individual ones + 1 ensemble) and pick the best ##
#####################################################################################

```

**Comment by mloning** [comment]
I understand what you're trying to do. I don't feel comfortable changing the ensemble class because I'm not sure what the overall design principle should be here. I'm wondering why we would change the ensemble class but not other composition classes like pipeline? Does scikit-learn have something similar? 

What you want to do is also already possible: you can also fit all models using ensemble and then retrieve the individual fitted models and see how they perform individually, no? 

On the other hand, it does seem to be a relatively common workflow to fit models separately and then try to combine them into ensembles without wanting to refit each of them.

**Comment by fkiraly** [comment]
>This is an interesting question, I'm not really sure what the best solution is. My intuition is that if we change how the ensemble works, we should also be open to change other composition classes like pipelines and I'm less sure about that.

My general opinion on the suggestion is that this comes at a substantial danger of information leakage, and "training on the test set via the modeller's brain".

In terms of architecture and user journey, I also see this negatively, since the behaviour of a specified modelling strategy may depend on the state of the objects passed ad construction (fitted or not). 

Conceptually, I think:
* construction should always create a modelling strategy in its initial state. It seems any deviation from this creates some inconsistencies in the behaviour that a concrete class of a given scitype models.
* specifically: calling `fit` for a model should always move it from the state "unfitted/initial" to "fitted". If you could construct a model partially fitted, the conceptual model for what `fit` does would be inconsistent. Therefore, allowing to pass partially fitted models at construction would violate the "strategy" pattern in this respect.

**Comment by ngupta23** [comment]
@fkiraly , Understand the concerns outlined above. Would you be open to providing an argument in the ensemble that does not fit an already fitted object (by default, it will maintain the current functionality of fitting again).

My concern is on the training time. If there are multiple models that are passed to the ensemble object and each of them have been fit using a Grid Search approach, it can take a long time and since the ensemble is currently only taking means, it is essentially duplicating the effort of fitting and doubling the training time.

**Comment by ngupta23** [comment]
@fkiraly 

> My general opinion on the suggestion is that this comes at a substantial danger of information leakage, and "training on the test set via the modeller's brain".

I can see where this would be an issue, especially when the ensemble is used with cross validation. Yes, that would be an issue for sure and may lead to leakage.

@mloning Sorry I missed your earlier message. 

> What you want to do is also already possible: you can also fit all models using ensemble and then retrieve the individual fitted models and see how they perform individually, no?

I think this is one option, but sometimes, you may only want to feed a subset of the individual models to the ensemble (only the top performing ones so as to not deteriorate the ensemble's performance), so the fitting all models using the ensemble, then pulling out the bad performing ones and refitting the good ones in the ensemble seems convoluted.

Given both your feedback, I am also not sure what the best approach would be. Let me give it some more thought. Maybe we can leave this open for some time.

**Comment by mloning** [comment]
Please feel free to reopen @ngupta23 if you have any more insights on this.
